// cive/_cjs/utils/typedData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeTypedData = serializeTypedData;
exports.validateTypedData = validateTypedData;
exports.getTypesForCIP23Domain = getTypesForCIP23Domain;
exports.domainSeparator = domainSeparator;
const viem_1 = require("viem");
const utils_1 = require("viem/utils");
const isAddress_js_1 = require("./address/isAddress.js");
const address_js_1 = require("./errors/address.js");
const hashTypedData_js_1 = require("./signature/hashTypedData.js");
function serializeTypedData(parameters) {
    const { domain: domain_, message: message_, primaryType, types, } = parameters;
    const normalizeData = (struct, data_) => {
        const data = { ...data_ };
        for (const param of struct) {
            const { name, type } = param;
            if (type === 'address')
                data[name] = data[name].toLowerCase();
        }
        return data;
    };
    const domain = (() => {
        if (!types.CIP23Domain)
            return {};
        if (!domain_)
            return {};
        return normalizeData(types.CIP23Domain, domain_);
    })();
    const message = (() => {
        if (primaryType === 'CIP23Domain')
            return undefined;
        return normalizeData(types[primaryType], message_);
    })();
    return (0, viem_1.stringify)({ domain, message, primaryType, types });
}
function validateTypedData(parameters) {
    const { domain, message, primaryType, types } = parameters;
    const validateData = (struct, data) => {
        for (const param of struct) {
            const { name, type } = param;
            const value = data[name];
            const integerMatch = type.match(utils_1.integerRegex);
            if (integerMatch &&
                (typeof value === 'number' || typeof value === 'bigint')) {
                const [_type, base, size_] = integerMatch;
                (0, viem_1.numberToHex)(value, {
                    signed: base === 'int',
                    size: Number.parseInt(size_) / 8,
                });
            }
            if (type === 'address' && typeof value === 'string' && !(0, isAddress_js_1.isAddress)(value))
                throw new address_js_1.InvalidAddressError({ address: value });
            const bytesMatch = type.match(utils_1.bytesRegex);
            if (bytesMatch) {
                const [_type, size_] = bytesMatch;
                if (size_ && (0, viem_1.size)(value) !== Number.parseInt(size_))
                    throw new viem_1.BytesSizeMismatchError({
                        expectedSize: Number.parseInt(size_),
                        givenSize: (0, viem_1.size)(value),
                    });
            }
            const struct = types[type];
            if (struct)
                validateData(struct, value);
        }
    };
    if (types.CIP23Domain && domain)
        validateData(types.CIP23Domain, domain);
    if (primaryType !== 'CIP23Domain')
        validateData(types[primaryType], message);
}
function getTypesForCIP23Domain({ domain, }) {
    return [
        typeof domain?.name === 'string' && { name: 'name', type: 'string' },
        domain?.version && { name: 'version', type: 'string' },
        typeof domain?.chainId === 'number' && {
            name: 'chainId',
            type: 'uint256',
        },
        domain?.verifyingContract && {
            name: 'verifyingContract',
            type: 'address',
        },
        domain?.salt && { name: 'salt', type: 'bytes32' },
    ].filter(Boolean);
}
function domainSeparator({ domain }) {
    return (0, hashTypedData_js_1.hashDomain)({
        domain,
        types: {
            CIP23Domain: getTypesForCIP23Domain({ domain }),
        },
    });
}
