// cive/_cjs/utils/formatters/account.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatAccountPending = formatAccountPending;
exports.formatAccountPendingTransaction = formatAccountPendingTransaction;
const transaction_js_1 = require("./transaction.js");
function formatAccountPending(accountPending) {
    const result = {
        ...accountPending,
        localNonce: accountPending.localNonce
            ? BigInt(accountPending.localNonce)
            : undefined,
        pendingNonce: accountPending.pendingNonce
            ? BigInt(accountPending.pendingNonce)
            : undefined,
        pendingCount: accountPending.pendingCount
            ? BigInt(accountPending.pendingCount)
            : undefined,
        nextPendingTx: accountPending.nextPendingTx
            ? BigInt(accountPending.nextPendingTx)
            : undefined,
    };
    return result;
}
function formatAccountPendingTransaction(accountPendingTransaction) {
    const result = {
        ...accountPendingTransaction,
        pendingCount: accountPendingTransaction.pendingCount
            ? BigInt(accountPendingTransaction.pendingCount)
            : undefined,
        pendingTransactions: accountPendingTransaction.pendingTransactions
            ? accountPendingTransaction.pendingTransactions.map(transaction_js_1.formatTransaction)
            : undefined,
    };
    return result;
}
