// cive/_cjs/utils/formatters/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transactionType = void 0;
exports.formatTransaction = formatTransaction;
const viem_1 = require("viem");
exports.transactionType = {
    '0x0': 'legacy',
    '0x1': 'eip2930',
    '0x2': 'eip1559',
};
function formatTransaction(transaction) {
    const _transaction = {
        ...transaction,
        blockHash: transaction.blockHash || null,
        chainId: transaction.chainId ? (0, viem_1.hexToNumber)(transaction.chainId) : undefined,
        contractCreated: transaction.contractCreated || null,
        epochHeight: transaction.epochHeight
            ? BigInt(transaction.epochHeight)
            : null,
        gas: BigInt(transaction.gas),
        gasPrice: transaction.gasPrice ? BigInt(transaction.gasPrice) : undefined,
        nonce: transaction.nonce ? Number(transaction.nonce) : undefined,
        status: transaction.status ? BigInt(transaction.status) : null,
        storageLimit: BigInt(transaction.storageLimit),
        to: transaction.to || null,
        transactionIndex: transaction.transactionIndex
            ? Number(transaction.transactionIndex)
            : null,
        v: BigInt(transaction.v),
        value: BigInt(transaction.value),
        maxFeePerGas: transaction.maxFeePerGas
            ? BigInt(transaction.maxFeePerGas)
            : undefined,
        maxPriorityFeePerGas: transaction.maxPriorityFeePerGas
            ? BigInt(transaction.maxPriorityFeePerGas)
            : undefined,
        type: transaction.type
            ? exports.transactionType[transaction.type]
            : undefined,
    };
    if (_transaction.type === 'legacy') {
        delete _transaction.accessList;
        delete _transaction.maxFeePerGas;
        delete _transaction.maxPriorityFeePerGas;
        delete _transaction.yParity;
    }
    if (_transaction.type === 'eip2930') {
        delete _transaction.maxFeePerGas;
        delete _transaction.maxPriorityFeePerGas;
    }
    return _transaction;
}
