// cive/_esm/actions/public/createEventFilter.js
import { encodeEventTopics, numberToHex } from 'viem/utils';
import { createFilterRequestScope } from '../../utils/filters/createFilterRequestScope.js';
export async function createEventFilter(client, { address, args, event, events: events_, fromBlock, strict, toBlock, fromEpoch, toEpoch, blockHashes, } = {}) {
    const events = events_ ?? (event ? [event] : undefined);
    const getRequest = createFilterRequestScope(client, {
        method: 'cfx_newFilter',
    });
    let topics = [];
    if (events) {
        const encoded = events.flatMap((event) => encodeEventTopics({
            abi: [event],
            eventName: event.name,
            args,
        }));
        // TODO: Clean up type casting
        topics = [encoded];
        if (event)
            topics = topics[0];
    }
    const fromBlock_ = typeof fromBlock === 'bigint' ? numberToHex(fromBlock) : fromBlock;
    const toBlock_ = typeof toBlock === 'bigint' ? numberToHex(toBlock) : toBlock;
    let id;
    if (blockHashes) {
        id = await client.request({
            method: 'cfx_newFilter',
            params: [
                {
                    fromBlock: fromBlock_,
                    toBlock: toBlock_,
                    blockHashes,
                    address,
                    ...(topics.length ? { topics } : {}),
                },
            ],
        });
    }
    else {
        const fromEpoch_ = typeof fromEpoch === 'bigint' ? numberToHex(fromEpoch) : fromEpoch;
        const toEpoch_ = typeof toEpoch === 'bigint' ? numberToHex(toEpoch) : toEpoch;
        id = await client.request({
            method: 'cfx_newFilter',
            params: [
                {
                    fromEpoch: fromEpoch_,
                    toEpoch: toEpoch_,
                    fromBlock: fromBlock_,
                    toBlock: toBlock_,
                    address,
                    ...(topics.length ? { topics } : {}),
                },
            ],
        });
    }
    return {
        abi: events,
        args,
        eventName: event ? event.name : undefined,
        fromBlock,
        id,
        request: getRequest(id),
        strict: Boolean(strict),
        toBlock,
        type: 'event',
    };
}
