// cive/_esm/actions/public/createContractEventFilter.js
import { encodeEventTopics, } from '../../utils/abi/encodeEventTopics.js';
import { numberToHex } from 'viem/utils';
import { createFilterRequestScope } from '../../utils/filters/createFilterRequestScope.js';
export async function createContractEventFilter(client, parameters) {
    const { address, abi, args, eventName, fromBlock, strict, toBlock, blockHashes, fromEpoch = 'latest_checkpoint', toEpoch = 'latest_state', } = parameters;
    const _fromEpoch = fromBlock ? undefined : fromEpoch;
    const _toEpoch = toBlock ? undefined : toEpoch;
    const getRequest = createFilterRequestScope(client, {
        method: 'cfx_newFilter',
    });
    const topics = eventName
        ? encodeEventTopics({
            abi,
            args,
            eventName,
        })
        : undefined;
    const params = blockHashes
        ? {
            blockHashes,
            address,
            topics,
        }
        : {
            fromEpoch: typeof _fromEpoch === 'bigint' ? numberToHex(_fromEpoch) : _fromEpoch,
            toEpoch: typeof _toEpoch === 'bigint' ? numberToHex(_toEpoch) : _toEpoch,
            fromBlock: typeof fromBlock === 'bigint' ? numberToHex(fromBlock) : fromBlock,
            toBlock: typeof toBlock === 'bigint' ? numberToHex(toBlock) : toBlock,
            address,
            topics,
        };
    const id = await client.request({
        method: 'cfx_newFilter',
        params: [params],
    });
    return {
        abi,
        args,
        eventName,
        id,
        request: getRequest(id),
        strict: Boolean(strict),
        type: 'event',
    };
}
