// cive/_esm/actions/public/getLogs.js
import { numberToHex } from 'viem/utils';
import { ChainIdNotFoundError } from '../../errors/chain.js';
import { encodeEventTopics, } from '../../utils/abi/encodeEventTopics.js';
import { parseEventLogs } from '../../utils/abi/parseEventLogs.js';
import { formatLog } from '../../utils/formatters/log.js';
export async function getLogs(client, { fromEpoch = 'latest_checkpoint', toEpoch = 'latest_state', fromBlock, toBlock, blockHashes, address, args, event, events: events_, strict: strict_, } = {}) {
    const strict = strict_ ?? false;
    const events = events_ ?? (event ? [event] : undefined);
    // TODO: update this
    if (typeof client.chain === 'undefined' || !('id' in client.chain)) {
        throw new ChainIdNotFoundError();
    }
    let topics = [];
    if (events) {
        const encoded = events.flatMap((event) => encodeEventTopics({
            abi: [event],
            eventName: event.name,
            args: events_ ? undefined : args,
        }));
        topics = [encoded];
        if (event)
            topics = topics[0];
    }
    const _fromEpoch = fromBlock ? undefined : fromEpoch;
    const _toEpoch = toBlock ? undefined : toEpoch;
    const params = blockHashes
        ? {
            blockHashes,
            address,
            topics,
        }
        : {
            fromEpoch: typeof _fromEpoch === 'bigint' ? numberToHex(_fromEpoch) : _fromEpoch,
            toEpoch: typeof _toEpoch === 'bigint' ? numberToHex(_toEpoch) : _toEpoch,
            fromBlock: typeof fromBlock === 'bigint' ? numberToHex(fromBlock) : fromBlock,
            toBlock: typeof toBlock === 'bigint' ? numberToHex(toBlock) : toBlock,
            address,
            topics,
        };
    const logs = await client.request({
        method: 'cfx_getLogs',
        params: [params],
    });
    const formattedLogs = logs.map((log) => formatLog(log));
    if (!events)
        return formattedLogs;
    return parseEventLogs({
        abi: events,
        logs: formattedLogs,
        strict,
        networkId: client.chain.id,
    });
}
