// cive/_esm/types/abitype/human-readable/errors/abiParameter.js
import { BaseError } from '../../errors.js';
export class InvalidAbiParameterError extends BaseError {
    name = 'InvalidAbiParameterError';
    constructor({ param }) {
        super('Failed to parse ABI parameter.', {
            details: `parseAbiParameter(${JSON.stringify(param, null, 2)})`,
            docsPath: '/api/human#parseabiparameter-1',
        });
    }
}
export class InvalidAbiParametersError extends BaseError {
    name = 'InvalidAbiParametersError';
    constructor({ params }) {
        super('Failed to parse ABI parameters.', {
            details: `parseAbiParameters(${JSON.stringify(params, null, 2)})`,
            docsPath: '/api/human#parseabiparameters-1',
        });
    }
}
export class InvalidParameterError extends BaseError {
    name = 'InvalidParameterError';
    constructor({ param }) {
        super('Invalid ABI parameter.', {
            details: param,
        });
    }
}
export class SolidityProtectedKeywordError extends BaseError {
    name = 'SolidityProtectedKeywordError';
    constructor({ param, name }) {
        super('Invalid ABI parameter.', {
            details: param,
            metaMessages: [
                `"${name}" is a protected Solidity keyword. More info: https://docs.soliditylang.org/en/latest/cheatsheet.html`,
            ],
        });
    }
}
export class InvalidModifierError extends BaseError {
    name = 'InvalidModifierError';
    constructor({ param, type, modifier, }) {
        super('Invalid ABI parameter.', {
            details: param,
            metaMessages: [
                `Modifier "${modifier}" not allowed${type ? ` in "${type}" type` : ''}.`,
            ],
        });
    }
}
export class InvalidFunctionModifierError extends BaseError {
    name = 'InvalidFunctionModifierError';
    constructor({ param, type, modifier, }) {
        super('Invalid ABI parameter.', {
            details: param,
            metaMessages: [
                `Modifier "${modifier}" not allowed${type ? ` in "${type}" type` : ''}.`,
                `Data location can only be specified for array, struct, or mapping types, but "${modifier}" was given.`,
            ],
        });
    }
}
export class InvalidAbiTypeParameterError extends BaseError {
    name = 'InvalidAbiTypeParameterError';
    constructor({ abiParameter, }) {
        super('Invalid ABI parameter.', {
            details: JSON.stringify(abiParameter, null, 2),
            metaMessages: ['ABI parameter type is invalid.'],
        });
    }
}
