// cive/_esm/utils/errors/chain.js
import { BaseError } from 'viem';
export class ChainDoesNotSupportContract extends BaseError {
    name = 'ChainDoesNotSupportContract';
    constructor({ epochNumber, chain, contract, }) {
        super(`Chain "${chain.name}" does not support contract "${contract.name}".`, {
            metaMessages: [
                'This could be due to any of the following:',
                ...(epochNumber &&
                    contract.blockCreated &&
                    contract.blockCreated > epochNumber
                    ? [
                        `- The contract "${contract.name}" was not deployed until block ${contract.blockCreated} (current block ${epochNumber}).`,
                    ]
                    : [
                        `- The chain does not have the contract "${contract.name}" configured.`,
                    ]),
            ],
        });
    }
}
export class ChainMismatchError extends BaseError {
    name = 'ChainMismatchError';
    constructor({ chain, currentChainId, }) {
        super(`The current chain of the wallet (id: ${currentChainId}) does not match the target chain for the transaction (id: ${chain.id} – ${chain.name}).`, {
            metaMessages: [
                `Current Chain ID:  ${currentChainId}`,
                `Expected Chain ID: ${chain.id} – ${chain.name}`,
            ],
        });
    }
}
export class ChainNotFoundError extends BaseError {
    name = 'ChainNotFoundError';
    constructor() {
        super([
            'No chain was provided to the request.',
            'Please provide a chain with the `chain` argument on the Action, or by supplying a `chain` to WalletClient.',
        ].join('\n'));
    }
}
export class ClientChainNotConfiguredError extends BaseError {
    name = 'ClientChainNotConfiguredError';
    constructor() {
        super('No chain was provided to the Client.');
    }
}
export class InvalidChainIdError extends BaseError {
    name = 'InvalidChainIdError';
    constructor({ chainId }) {
        super(typeof chainId === 'number' || typeof chainId === 'string'
            ? `Chain ID "${chainId}" is invalid.`
            : 'Chain ID is invalid.');
    }
}
export class InvalidChainNameError extends BaseError {
    name = 'InvalidChainNameError';
    constructor({ chainName }) {
        super(typeof chainName === 'string'
            ? `Chain name "${chainName}" is invalid.should be passed by 'cfx', 'cfxtest' or 'net[n], net1 or net1029 are invalid`
            : 'Chain name is invalid.');
    }
}
