// cive/_esm/actions/wallet/prepareTransactionRequest.js
import { Eip1559FeesNotSupportedError, MaxFeePerGasTooLowError, } from 'viem';
import { parseAccount, } from '../../accounts/index.js';
import { transactionGas, transactionStorageLimit, } from '../../constants/transaction.js';
import { decodeBase32Address } from '../../utils/address/decodeBase32Address.js';
import { getAction } from '../../utils/getAction.js';
import { assertRequest, } from '../../utils/transaction/assertRequest.js';
import { getTransactionType, } from '../../utils/transaction/getTransactionType.js';
import { internal_estimateFeesPerGas } from '../public/estimateFeesPerGas.js';
import { estimateGasAndCollateral, } from '../public/estimateGasAndCollateral.js';
import { getBlock as getBlock_ } from '../public/getBlock.js';
import { getEpochNumber } from '../public/getEpochNumber.js';
import { getNextNonce } from '../public/getNextNonce.js';
import { getStatus } from '../public/getStatus.js';
export const defaultParameters = [
    'chainId',
    'fees',
    'gas',
    'nonce',
    'type',
    'epochHeight',
    'storageLimit',
];
export async function prepareTransactionRequest(client, args) {
    const { account: account_ = client.account, chain, chainId, gas, nonce, parameters = defaultParameters, type, epochHeight, storageLimit, to, data, accessList, } = args;
    const account = account_ ? parseAccount(account_) : undefined;
    const request = {
        ...args,
        ...(account ? { from: account?.address } : {}),
        to,
        data,
        accessList,
    };
    let block;
    async function getBlock(epochNumber) {
        if (block)
            return block;
        block = await getAction(client, getBlock_, 'getBlock')({ epochNumber: epochNumber });
        return block;
    }
    if (parameters.includes('nonce') && typeof nonce === 'undefined' && account) {
        /**
         * The execution of conflux transactions will be delayed for several seconds.
         * During this period, the nonce obtained by getNextNonce will still be old,
         * while txPoolNextNonce will provide the correct nonce.
         * but not all RPC nodes support txPoolNextNonce.
         */
        request.nonce = await getAction(client, getNextNonce, 'getNextNonce')({
            address: account.address,
        });
    }
    if (parameters.includes('chainId')) {
        if (chain) {
            request.chainId = chain.id;
        }
        else if (typeof chainId !== 'undefined') {
            request.chainId = chainId;
        }
        else {
            const nodeStatus = await getAction(client, getStatus, 'getChainId')({});
            request.chainId = nodeStatus.chainId;
        }
    }
    if (parameters.includes('epochHeight')) {
        if (typeof epochHeight !== 'undefined') {
            request.epochHeight = epochHeight;
        }
        else {
            request.epochHeight = await getAction(client, getEpochNumber, 'getEpochNumber')({ cacheTime: 0 });
        }
    }
    if ((parameters.includes('fees') || parameters.includes('type')) &&
        typeof type === 'undefined') {
        try {
            request.type = getTransactionType(request);
        }
        catch {
            // infer type from block
            const block = await getBlock(request.epochHeight);
            request.type =
                typeof block?.baseFeePerGas === 'bigint' ? 'eip1559' : 'legacy';
        }
    }
    if ((parameters.includes('gas') && typeof gas === 'undefined') ||
        (parameters.includes('storageLimit') && typeof storageLimit === 'undefined')) {
        const isToUser = to && decodeBase32Address({ address: to }).type === 'user';
        if (isToUser && !data && !accessList) {
            request.gas = transactionGas;
            request.storageLimit = transactionStorageLimit;
        }
        else {
            const { gasLimit, storageCollateralized } = await getAction(client, estimateGasAndCollateral, 'estimateGasAndCollateral')({
                ...request,
                account: account
                    ? { address: account.address, type: 'json-rpc' }
                    : undefined,
            });
            if (typeof gas === 'undefined') {
                request.gas = gasLimit;
            }
            if (typeof storageLimit === 'undefined') {
                request.storageLimit = storageCollateralized;
            }
        }
    }
    if (parameters.includes('fees')) {
        if (request.type !== 'legacy' && request.type !== 'eip2930') {
            // EIP-1559 fees
            if (typeof request.maxFeePerGas === 'undefined' ||
                typeof request.maxPriorityFeePerGas === 'undefined') {
                const block = await getBlock(request.epochHeight);
                const { maxFeePerGas, maxPriorityFeePerGas } = await internal_estimateFeesPerGas(client, {
                    block: block,
                    chain,
                    request: request,
                });
                if (typeof args.maxPriorityFeePerGas === 'undefined' &&
                    args.maxFeePerGas &&
                    args.maxFeePerGas < maxPriorityFeePerGas)
                    throw new MaxFeePerGasTooLowError({
                        maxPriorityFeePerGas,
                    });
                request.maxPriorityFeePerGas = maxPriorityFeePerGas;
                request.maxFeePerGas = maxFeePerGas;
            }
        }
        else {
            // Legacy fees
            if (typeof args.maxFeePerGas !== 'undefined' ||
                typeof args.maxPriorityFeePerGas !== 'undefined')
                throw new Eip1559FeesNotSupportedError();
            const block = await getBlock(request.epochHeight);
            const { gasPrice: gasPrice_ } = await internal_estimateFeesPerGas(client, {
                block: block,
                chain,
                request: request,
                type: 'legacy',
            });
            request.gasPrice = gasPrice_;
        }
    }
    assertRequest(request);
    delete request.parameters;
    return request;
}
