// cive/_cjs/types/abitype/human-readable/errors/abiItem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnknownSolidityTypeError = exports.UnknownTypeError = exports.InvalidAbiItemError = void 0;
const errors_js_1 = require("../../errors.js");
class InvalidAbiItemError extends errors_js_1.BaseError {
    name = 'InvalidAbiItemError';
    constructor({ signature }) {
        super('Failed to parse ABI item.', {
            details: `parseAbiItem(${JSON.stringify(signature, null, 2)})`,
            docsPath: '/api/human#parseabiitem-1',
        });
    }
}
exports.InvalidAbiItemError = InvalidAbiItemError;
class UnknownTypeError extends errors_js_1.BaseError {
    name = 'UnknownTypeError';
    constructor({ type }) {
        super('Unknown type.', {
            metaMessages: [
                `Type "${type}" is not a valid ABI type. Perhaps you forgot to include a struct signature?`,
            ],
        });
    }
}
exports.UnknownTypeError = UnknownTypeError;
class UnknownSolidityTypeError extends errors_js_1.BaseError {
    name = 'UnknownSolidityTypeError';
    constructor({ type }) {
        super('Unknown type.', {
            metaMessages: [`Type "${type}" is not a valid ABI type.`],
        });
    }
}
exports.UnknownSolidityTypeError = UnknownSolidityTypeError;
