// cive/_esm/utils/formatters/account.js
import { formatTransaction } from './transaction.js';
export function formatAccountPending(accountPending) {
    const result = {
        ...accountPending,
        localNonce: accountPending.localNonce
            ? BigInt(accountPending.localNonce)
            : undefined,
        pendingNonce: accountPending.pendingNonce
            ? BigInt(accountPending.pendingNonce)
            : undefined,
        pendingCount: accountPending.pendingCount
            ? BigInt(accountPending.pendingCount)
            : undefined,
        nextPendingTx: accountPending.nextPendingTx
            ? BigInt(accountPending.nextPendingTx)
            : undefined,
    };
    return result;
}
export function formatAccountPendingTransaction(accountPendingTransaction) {
    const result = {
        ...accountPendingTransaction,
        pendingCount: accountPendingTransaction.pendingCount
            ? BigInt(accountPendingTransaction.pendingCount)
            : undefined,
        pendingTransactions: accountPendingTransaction.pendingTransactions
            ? accountPendingTransaction.pendingTransactions.map(formatTransaction)
            : undefined,
    };
    return result;
}
