// cive/_esm/utils/formatters/block.js
import { formatTransaction } from './transaction.js';
export function formatBlock(block) {
    const transactions = block.transactions?.map((transaction) => {
        if (typeof transaction === 'string')
            return transaction;
        return formatTransaction(transaction);
    });
    const _block = {
        ...block,
        baseFeePerGas: block.baseFeePerGas ? BigInt(block.baseFeePerGas) : null,
        blame: block.blame ? BigInt(block.blame) : undefined,
        difficulty: block.difficulty ? BigInt(block.difficulty) : undefined,
        epochNumber: block.epochNumber ? BigInt(block.epochNumber) : undefined,
        gasLimit: block.gasLimit ? BigInt(block.gasLimit) : undefined,
        gasUsed: block.gasUsed ? BigInt(block.gasUsed) : null,
        height: block.height ? BigInt(block.height) : undefined,
        size: block.size ? BigInt(block.size) : undefined,
        timestamp: block.timestamp ? BigInt(block.timestamp) : undefined,
        blockNumber: block.blockNumber ? BigInt(block.blockNumber) : null,
        transactions,
    };
    return _block;
}
