// cive/_esm/utils/formatters/tract.js
export function formatTract(trace) {
    const result = {
        ...trace,
        epochNumber: trace?.epochNumber ? BigInt(trace.epochNumber) : undefined,
        transactionPosition: trace?.transactionPosition
            ? BigInt(trace.transactionPosition)
            : undefined,
    };
    if ('value' in result.action) {
        result.action.value = BigInt(result.action.value);
    }
    if ('gas' in result.action) {
        result.action.gas = BigInt(result.action.gas);
    }
    if ('gasLeft' in result.action) {
        result.action.gasLeft = BigInt(result.action.gasLeft);
    }
    return result;
}
export function formatTransactionTraces(transactionTraces) {
    const result = {
        ...transactionTraces,
        transactionPosition: transactionTraces?.transactionPosition
            ? BigInt(transactionTraces.transactionPosition)
            : undefined,
        traces: transactionTraces?.traces
            ? transactionTraces.traces.map(formatTract)
            : undefined,
    };
    return result;
}
export function formatTraceBlock(block) {
    const result = {
        ...block,
        epochNumber: block?.epochNumber ? BigInt(block.epochNumber) : undefined,
        transactionTraces: block?.transactionTraces
            ? block.transactionTraces.map(formatTransactionTraces)
            : undefined,
    };
    return result;
}
