// cive/_cjs/actions/public/createContractEventFilter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createContractEventFilter = createContractEventFilter;
const encodeEventTopics_js_1 = require("../../utils/abi/encodeEventTopics.js");
const utils_1 = require("viem/utils");
const createFilterRequestScope_js_1 = require("../../utils/filters/createFilterRequestScope.js");
async function createContractEventFilter(client, parameters) {
    const { address, abi, args, eventName, fromBlock, strict, toBlock, blockHashes, fromEpoch = 'latest_checkpoint', toEpoch = 'latest_state', } = parameters;
    const _fromEpoch = fromBlock ? undefined : fromEpoch;
    const _toEpoch = toBlock ? undefined : toEpoch;
    const getRequest = (0, createFilterRequestScope_js_1.createFilterRequestScope)(client, {
        method: 'cfx_newFilter',
    });
    const topics = eventName
        ? (0, encodeEventTopics_js_1.encodeEventTopics)({
            abi,
            args,
            eventName,
        })
        : undefined;
    const params = blockHashes
        ? {
            blockHashes,
            address,
            topics,
        }
        : {
            fromEpoch: typeof _fromEpoch === 'bigint' ? (0, utils_1.numberToHex)(_fromEpoch) : _fromEpoch,
            toEpoch: typeof _toEpoch === 'bigint' ? (0, utils_1.numberToHex)(_toEpoch) : _toEpoch,
            fromBlock: typeof fromBlock === 'bigint' ? (0, utils_1.numberToHex)(fromBlock) : fromBlock,
            toBlock: typeof toBlock === 'bigint' ? (0, utils_1.numberToHex)(toBlock) : toBlock,
            address,
            topics,
        };
    const id = await client.request({
        method: 'cfx_newFilter',
        params: [params],
    });
    return {
        abi,
        args,
        eventName,
        id,
        request: getRequest(id),
        strict: Boolean(strict),
        type: 'event',
    };
}
