// cive/_cjs/actions/public/getFilterChanges.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFilterChanges = getFilterChanges;
const chain_js_1 = require("../../errors/chain.js");
const parseEventLogs_js_1 = require("../../utils/abi/parseEventLogs.js");
const log_js_1 = require("../../utils/formatters/log.js");
async function getFilterChanges(client, { filter, }) {
    const strict = 'strict' in filter && filter.strict;
    if (typeof client.chain === 'undefined' || !('id' in client.chain)) {
        throw new chain_js_1.ChainIdNotFoundError();
    }
    const logs = await filter.request({
        method: 'cfx_getFilterChanges',
        params: [filter.id],
    });
    if (typeof logs[0] === 'string')
        return logs;
    const formattedLogs = logs.map((log) => (0, log_js_1.formatLog)(log));
    if (!('abi' in filter) || !filter.abi)
        return formattedLogs;
    return (0, parseEventLogs_js_1.parseEventLogs)({
        abi: filter.abi,
        logs: formattedLogs,
        strict,
        networkId: client.chain.id,
    });
}
