// cive/_cjs/actions/public/verifyHash.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifyHash = verifyHash;
const utils_1 = require("viem/utils");
const index_js_1 = require("../../accounts/index.js");
const abis_js_1 = require("../../constants/abis.js");
const contract_js_1 = require("../../constants/contract.js");
const contract_js_2 = require("../../errors/contract.js");
const isBytesEqual_js_1 = require("../../utils/data/isBytesEqual.js");
const getAction_js_1 = require("../../utils/getAction.js");
const index_js_2 = require("../../utils/index.js");
const serializeErc6492Signature_js_1 = require("../../utils/signature/serializeErc6492Signature.js");
const call_js_1 = require("./call.js");
async function verifyHash(client, parameters) {
    const { address, factory, factoryData, hash, signature, ...rest } = parameters;
    const signatureHex = (() => {
        if ((0, utils_1.isHex)(signature))
            return signature;
        if (typeof signature === 'object' && 'r' in signature && 's' in signature)
            return (0, index_js_1.serializeSignature)(signature);
        return (0, utils_1.bytesToHex)(signature);
    })();
    const wrappedSignature = await (async () => {
        if (!factory && !factoryData)
            return signatureHex;
        if ((0, utils_1.isErc6492Signature)(signatureHex))
            return signatureHex;
        return (0, serializeErc6492Signature_js_1.serializeErc6492Signature)({
            address: factory,
            data: factoryData,
            signature: signatureHex,
        });
    })();
    try {
        const { data } = await (0, getAction_js_1.getAction)(client, call_js_1.call, 'call')({
            data: (0, index_js_2.encodeDeployData)({
                abi: abis_js_1.universalSignatureValidatorAbi,
                args: [address, hash, wrappedSignature],
                bytecode: contract_js_1.universalSignatureValidatorByteCode,
            }),
            ...rest,
        });
        return (0, isBytesEqual_js_1.isBytesEqual)(data ?? '0x0', '0x1');
    }
    catch (error) {
        try {
            const verified = (0, utils_1.isAddressEqual)((0, index_js_2.base32AddressToHex)({ address: (0, index_js_2.getAddress)(address) }), await (0, utils_1.recoverAddress)({ hash, signature }));
            if (verified)
                return true;
        }
        catch { }
        if (error instanceof contract_js_2.CallExecutionError) {
            return false;
        }
        throw error;
    }
}
