// cive/_cjs/clients/decorators/public.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.publicActions = publicActions;
const verifyTypedData_js_1 = require("../../actions/public//verifyTypedData.js");
const call_js_1 = require("../../actions/public/call.js");
const checkBalanceAgainstTransaction_js_1 = require("../../actions/public/checkBalanceAgainstTransaction.js");
const createBlockFilter_js_1 = require("../../actions/public/createBlockFilter.js");
const createContractEventFilter_js_1 = require("../../actions/public/createContractEventFilter.js");
const createEventFilter_js_1 = require("../../actions/public/createEventFilter.js");
const createPendingTransactionFilter_js_1 = require("../../actions/public/createPendingTransactionFilter.js");
const estimateContractGasAndCollateral_js_1 = require("../../actions/public/estimateContractGasAndCollateral.js");
const estimateGasAndCollateral_js_1 = require("../../actions/public/estimateGasAndCollateral.js");
const estimateMaxPriorityFeePerGas_js_1 = require("../../actions/public/estimateMaxPriorityFeePerGas.js");
const getAccount_js_1 = require("../../actions/public/getAccount.js");
const getAccountPendingInfo_js_1 = require("../../actions/public/getAccountPendingInfo.js");
const getAccountPendingTransactions_js_1 = require("../../actions/public/getAccountPendingTransactions.js");
const getAccumulateInterestRate_js_1 = require("../../actions/public/getAccumulateInterestRate.js");
const getAdmin_js_1 = require("../../actions/public/getAdmin.js");
const getBalance_js_1 = require("../../actions/public/getBalance.js");
const getBastBlockHash_js_1 = require("../../actions/public/getBastBlockHash.js");
const getBlock_js_1 = require("../../actions/public/getBlock.js");
const getBlockByHashWithPivotAssumption_js_1 = require("../../actions/public/getBlockByHashWithPivotAssumption.js");
const getBlockRewardInfo_js_1 = require("../../actions/public/getBlockRewardInfo.js");
const getBlocksByEpoch_js_1 = require("../../actions/public/getBlocksByEpoch.js");
const getChainId_js_1 = require("../../actions/public/getChainId.js");
const getClientVersion_js_1 = require("../../actions/public/getClientVersion.js");
const getCode_js_1 = require("../../actions/public/getCode.js");
const getCollateralInfo_js_1 = require("../../actions/public/getCollateralInfo.js");
const getCollaterlForStorage_js_1 = require("../../actions/public/getCollaterlForStorage.js");
const getConfirmationRiskByHash_js_1 = require("../../actions/public/getConfirmationRiskByHash.js");
const getContractEvents_js_1 = require("../../actions/public/getContractEvents.js");
const getDepositList_js_1 = require("../../actions/public/getDepositList.js");
const getEpochNumber_js_1 = require("../../actions/public/getEpochNumber.js");
const getEpochReceipts_js_1 = require("../../actions/public/getEpochReceipts.js");
const getFeeBurnt_js_1 = require("../../actions/public/getFeeBurnt.js");
const getFeeHistory_js_1 = require("../../actions/public/getFeeHistory.js");
const getFilterChanges_js_1 = require("../../actions/public/getFilterChanges.js");
const getFilterLogs_js_1 = require("../../actions/public/getFilterLogs.js");
const getGasPrice_js_1 = require("../../actions/public/getGasPrice.js");
const getInterestRate_js_1 = require("../../actions/public/getInterestRate.js");
const getLogs_js_1 = require("../../actions/public/getLogs.js");
const getNextNonce_js_1 = require("../../actions/public/getNextNonce.js");
const getParamsFromVote_js_1 = require("../../actions/public/getParamsFromVote.js");
const getPoSAccount_js_1 = require("../../actions/public/getPoSAccount.js");
const getPoSBlock_js_1 = require("../../actions/public/getPoSBlock.js");
const getPoSCommittee_js_1 = require("../../actions/public/getPoSCommittee.js");
const getPoSEconomics_js_1 = require("../../actions/public/getPoSEconomics.js");
const getPoSRewardByEpoch_js_1 = require("../../actions/public/getPoSRewardByEpoch.js");
const getPoSRewards_js_1 = require("../../actions/public/getPoSRewards.js");
const getPoSStatus_js_1 = require("../../actions/public/getPoSStatus.js");
const getPoSTransactionByNumber_js_1 = require("../../actions/public/getPoSTransactionByNumber.js");
const getSkippedBlocksByEpoch_js_1 = require("../../actions/public/getSkippedBlocksByEpoch.js");
const getSponsorInfo_js_1 = require("../../actions/public/getSponsorInfo.js");
const getStakingBalance_js_1 = require("../../actions/public/getStakingBalance.js");
const getStatus_js_1 = require("../../actions/public/getStatus.js");
const getStorageAt_js_1 = require("../../actions/public/getStorageAt.js");
const getStorageRoot_js_1 = require("../../actions/public/getStorageRoot.js");
const getSupplyInfo_js_1 = require("../../actions/public/getSupplyInfo.js");
const getTransaction_js_1 = require("../../actions/public/getTransaction.js");
const getTransactionReceipt_js_1 = require("../../actions/public/getTransactionReceipt.js");
const getTxPoolNextNonce_js_1 = require("../../actions/public/getTxPoolNextNonce.js");
const getVoteList_js_1 = require("../../actions/public/getVoteList.js");
const multicall_js_1 = require("../../actions/public/multicall.js");
const readContract_js_1 = require("../../actions/public/readContract.js");
const simulateContract_js_1 = require("../../actions/public/simulateContract.js");
const traceBlock_js_1 = require("../../actions/public/traceBlock.js");
const traceTransaction_js_1 = require("../../actions/public/traceTransaction.js");
const uninstallFilter_js_1 = require("../../actions/public/uninstallFilter.js");
const verifyMessage_js_1 = require("../../actions/public/verifyMessage.js");
const waitForTransactionReceipt_js_1 = require("../../actions/public/waitForTransactionReceipt.js");
const watchEpochNumber_js_1 = require("../../actions/public/watchEpochNumber.js");
const sendRawTransaction_js_1 = require("../../actions/wallet/sendRawTransaction.js");
function publicActions(client) {
    return {
        getTransaction: (args) => (0, getTransaction_js_1.getTransaction)(client, args),
        getBlock: (args) => (0, getBlock_js_1.getBlock)(client, args),
        getBastBlockHash: () => (0, getBastBlockHash_js_1.getBastBlockHash)(client),
        getEpochNumber: (args) => (0, getEpochNumber_js_1.getEpochNumber)(client, args),
        getGasPrice: () => (0, getGasPrice_js_1.getGasPrice)(client),
        estimateMaxPriorityFeePerGas: () => (0, estimateMaxPriorityFeePerGas_js_1.estimateMaxPriorityFeePerGas)(client),
        getFeeHistory: (args) => (0, getFeeHistory_js_1.getFeeHistory)(client, args),
        getBlocksByEpoch: (args) => (0, getBlocksByEpoch_js_1.getBlocksByEpoch)(client, args),
        getBalance: (args) => (0, getBalance_js_1.getBalance)(client, args),
        getStakingBalance: (args) => (0, getStakingBalance_js_1.getStakingBalance)(client, args),
        getCollateralForStorage: (args) => (0, getCollaterlForStorage_js_1.getCollateralForStorage)(client, args),
        getAdmin: (args) => (0, getAdmin_js_1.getAdmin)(client, args),
        getCode: (args) => (0, getCode_js_1.getCode)(client, args),
        getStorageAt: (args) => (0, getStorageAt_js_1.GetStorageAt)(client, args),
        getStorageRoot: (args) => (0, getStorageRoot_js_1.GetStorageRoot)(client, args),
        getSponsorInfo: (args) => (0, getSponsorInfo_js_1.GetSponsorInfo)(client, args),
        getNextNonce: (args) => (0, getNextNonce_js_1.getNextNonce)(client, args),
        call: (args) => (0, call_js_1.call)(client, args),
        estimateGasAndCollateral: (args) => (0, estimateGasAndCollateral_js_1.estimateGasAndCollateral)(client, args),
        getLogs: (args) => (0, getLogs_js_1.getLogs)(client, args),
        getTransactionReceipt: (args) => (0, getTransactionReceipt_js_1.getTransactionReceipt)(client, args),
        getAccount: (args) => (0, getAccount_js_1.getAccount)(client, args),
        getInterestRate: (args) => (0, getInterestRate_js_1.getInterestRate)(client, args),
        getAccumulateInterestRate: (args) => (0, getAccumulateInterestRate_js_1.getAccumulateInterestRate)(client, args),
        checkBalanceAgainstTransaction: (args) => (0, checkBalanceAgainstTransaction_js_1.checkBalanceAgainstTransaction)(client, args),
        getSkippedBlocksByEpoch: (args) => (0, getSkippedBlocksByEpoch_js_1.getSkippedBlocksByEpoch)(client, args),
        getConfirmationRiskByHash: (args) => (0, getConfirmationRiskByHash_js_1.getConfirmationRiskByHash)(client, args),
        getStatus: () => (0, getStatus_js_1.getStatus)(client),
        getClientVersion: () => (0, getClientVersion_js_1.getClientVersion)(client),
        getBlockRewardInfo: (args) => (0, getBlockRewardInfo_js_1.getBlockRewardInfo)(client, args),
        getBlockByHashWithPivotAssumption: (args) => (0, getBlockByHashWithPivotAssumption_js_1.getBlockByHashWithPivotAssumption)(client, args),
        getDepositList: (args) => (0, getDepositList_js_1.getDepositList)(client, args),
        getVoteList: (args) => (0, getVoteList_js_1.getVoteList)(client, args),
        getSupplyInfo: () => (0, getSupplyInfo_js_1.getSupplyInfo)(client),
        getAccountPendingInfo: (args) => (0, getAccountPendingInfo_js_1.getAccountPendingInfo)(client, args),
        getAccountPendingTransactions: (args) => (0, getAccountPendingTransactions_js_1.getAccountPendingTransactions)(client, args),
        getPoSEconomics: (args) => (0, getPoSEconomics_js_1.getPoSEconomics)(client, args),
        getPoSRewardByEpoch: (args) => (0, getPoSRewardByEpoch_js_1.getPoSRewardByEpoch)(client, args),
        getParamsFromVote: (args) => (0, getParamsFromVote_js_1.getParamsFromVote)(client, args),
        createEventFilter: (args) => (0, createEventFilter_js_1.createEventFilter)(client, args),
        createBlockFilter: () => (0, createBlockFilter_js_1.createBlockFilter)(client),
        createPendingTransactionFilter: () => (0, createPendingTransactionFilter_js_1.createPendingTransactionFilter)(client),
        getFilterChanges: (args) => (0, getFilterChanges_js_1.getFilterChanges)(client, args),
        getFilterLogs: (args) => (0, getFilterLogs_js_1.getFilterLogs)(client, args),
        uninstallFilter: (args) => (0, uninstallFilter_js_1.uninstallFilter)(client, args),
        getCollateralInfo: (args) => (0, getCollateralInfo_js_1.getCollateralInfo)(client, args),
        getTxPoolNextNonce: (args) => (0, getTxPoolNextNonce_js_1.getTxPoolNextNonce)(client, args),
        getPoSStatus: () => (0, getPoSStatus_js_1.getPoSStatus)(client),
        getPoSAccount: (args) => (0, getPoSAccount_js_1.getPoSAccount)(client, args),
        getPoSCommittee: (args) => (0, getPoSCommittee_js_1.getPoSCommittee)(client, args),
        getPoSBlock: (args) => (0, getPoSBlock_js_1.getPoSBlock)(client, args),
        getPoSRewards: (args) => (0, getPoSRewards_js_1.getPoSRewards)(client, args),
        getPoSTransactionByNumber: (args) => (0, getPoSTransactionByNumber_js_1.getPoSTransactionByNumber)(client, args),
        getEpochReceipts: (args) => (0, getEpochReceipts_js_1.getEpochReceipts)(client, args),
        traceBlock: (args) => (0, traceBlock_js_1.traceBlock)(client, args),
        traceTransaction: (args) => (0, traceTransaction_js_1.traceTransaction)(client, args),
        getChainId: () => (0, getChainId_js_1.getChainId)(client),
        sendRawTransaction: (args) => (0, sendRawTransaction_js_1.sendRawTransaction)(client, args),
        getFeeBurnt: () => (0, getFeeBurnt_js_1.getFeeBurnt)(client),
        createContractEventFilter: (args) => (0, createContractEventFilter_js_1.createContractEventFilter)(client, args),
        readContract: (args) => (0, readContract_js_1.readContract)(client, args),
        simulateContract: (args) => (0, simulateContract_js_1.simulateContract)(client, args),
        getContractEvents: (args) => (0, getContractEvents_js_1.getContractEvents)(client, args),
        estimateContractGasAndCollateral: (args) => (0, estimateContractGasAndCollateral_js_1.estimateContractGasAndCollateral)(client, args),
        watchEpochNumber: (args) => (0, watchEpochNumber_js_1.watchEpochNumber)(client, args),
        waitForTransactionReceipt: (args) => (0, waitForTransactionReceipt_js_1.waitForTransactionReceipt)(client, args),
        multicall: (args) => (0, multicall_js_1.multicall)(client, args),
        verifyMessage: (args) => (0, verifyMessage_js_1.verifyMessage)(client, args),
        verifyTypedData: (args) => (0, verifyTypedData_js_1.verifyTypedData)(client, args),
    };
}
