// cive/_cjs/clients/decorators/test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.testActions = testActions;
const index_js_1 = require("../../actions/index.js");
const clearTxpool_js_1 = require("../../actions/test/clearTxpool.js");
const createLocalNodeAccount_js_1 = require("../../actions/test/createLocalNodeAccount.js");
const generateEmptyLocalNodeBlocks_js_1 = require("../../actions/test/generateEmptyLocalNodeBlocks.js");
const generateLocalNodeBlock_js_1 = require("../../actions/test/generateLocalNodeBlock.js");
const getLocalNodeAddresses_js_1 = require("../../actions/test/getLocalNodeAddresses.js");
const lockLocalNodeAccount_js_1 = require("../../actions/test/lockLocalNodeAccount.js");
const mine_js_1 = require("../../actions/test/mine.js");
const unlockLocalNodeAccount_js_1 = require("../../actions/test/unlockLocalNodeAccount.js");
function testActions(client) {
    return {
        clearTxpool: () => (0, clearTxpool_js_1.clearTxpool)(client),
        getLocalNodeAddresses: () => (0, getLocalNodeAddresses_js_1.getLocalNodeAddresses)(client),
        createLocalNodeAccount: (args) => (0, createLocalNodeAccount_js_1.createLocalNodeAccount)(client, args),
        unlockLocalNodeAccount: (args) => (0, unlockLocalNodeAccount_js_1.unlockLocalNodeAccount)(client, args),
        lockLocalNodeAccount: (args) => (0, lockLocalNodeAccount_js_1.lockLocalNodeAccount)(client, args),
        generateLocalNodeBlock: (args) => (0, generateLocalNodeBlock_js_1.generateLocalNodeBlock)(client, args),
        generateEmptyLocalNodeBlocks: (args) => (0, generateEmptyLocalNodeBlocks_js_1.generateEmptyLocalNodeBlocks)(client, args),
        mine: (args) => (0, mine_js_1.mine)(client, args),
        getCurrentSyncPhase: () => (0, index_js_1.getCurrentSyncPhase)(client),
    };
}
