// cive/_esm/clients/decorators/public.js
import { verifyTypedData, } from '../../actions/public//verifyTypedData.js';
import { call, } from '../../actions/public/call.js';
import { checkBalanceAgainstTransaction, } from '../../actions/public/checkBalanceAgainstTransaction.js';
import { createBlockFilter, } from '../../actions/public/createBlockFilter.js';
import { createContractEventFilter, } from '../../actions/public/createContractEventFilter.js';
import { createEventFilter, } from '../../actions/public/createEventFilter.js';
import { createPendingTransactionFilter, } from '../../actions/public/createPendingTransactionFilter.js';
import { estimateContractGasAndCollateral, } from '../../actions/public/estimateContractGasAndCollateral.js';
import { estimateGasAndCollateral, } from '../../actions/public/estimateGasAndCollateral.js';
import { estimateMaxPriorityFeePerGas, } from '../../actions/public/estimateMaxPriorityFeePerGas.js';
import { getAccount, } from '../../actions/public/getAccount.js';
import { getAccountPendingInfo, } from '../../actions/public/getAccountPendingInfo.js';
import { getAccountPendingTransactions, } from '../../actions/public/getAccountPendingTransactions.js';
import { getAccumulateInterestRate, } from '../../actions/public/getAccumulateInterestRate.js';
import { getAdmin, } from '../../actions/public/getAdmin.js';
import { getBalance, } from '../../actions/public/getBalance.js';
import { getBastBlockHash, } from '../../actions/public/getBastBlockHash.js';
import { getBlock, } from '../../actions/public/getBlock.js';
import { getBlockByHashWithPivotAssumption, } from '../../actions/public/getBlockByHashWithPivotAssumption.js';
import { getBlockRewardInfo, } from '../../actions/public/getBlockRewardInfo.js';
import { getBlocksByEpoch, } from '../../actions/public/getBlocksByEpoch.js';
import { getChainId, } from '../../actions/public/getChainId.js';
import { getClientVersion, } from '../../actions/public/getClientVersion.js';
import { getCode, } from '../../actions/public/getCode.js';
import { getCollateralInfo, } from '../../actions/public/getCollateralInfo.js';
import { getCollateralForStorage, } from '../../actions/public/getCollaterlForStorage.js';
import { getConfirmationRiskByHash, } from '../../actions/public/getConfirmationRiskByHash.js';
import { getContractEvents, } from '../../actions/public/getContractEvents.js';
import { getDepositList, } from '../../actions/public/getDepositList.js';
import { getEpochNumber, } from '../../actions/public/getEpochNumber.js';
import { getEpochReceipts, } from '../../actions/public/getEpochReceipts.js';
import { getFeeBurnt, } from '../../actions/public/getFeeBurnt.js';
import { getFeeHistory, } from '../../actions/public/getFeeHistory.js';
import { getFilterChanges, } from '../../actions/public/getFilterChanges.js';
import { getFilterLogs, } from '../../actions/public/getFilterLogs.js';
import { getGasPrice, } from '../../actions/public/getGasPrice.js';
import { getInterestRate, } from '../../actions/public/getInterestRate.js';
import { getLogs, } from '../../actions/public/getLogs.js';
import { getNextNonce, } from '../../actions/public/getNextNonce.js';
import { getParamsFromVote, } from '../../actions/public/getParamsFromVote.js';
import { getPoSAccount, } from '../../actions/public/getPoSAccount.js';
import { getPoSBlock, } from '../../actions/public/getPoSBlock.js';
import { getPoSCommittee, } from '../../actions/public/getPoSCommittee.js';
import { getPoSEconomics, } from '../../actions/public/getPoSEconomics.js';
import { getPoSRewardByEpoch, } from '../../actions/public/getPoSRewardByEpoch.js';
import { getPoSRewards, } from '../../actions/public/getPoSRewards.js';
import { getPoSStatus, } from '../../actions/public/getPoSStatus.js';
import { getPoSTransactionByNumber, } from '../../actions/public/getPoSTransactionByNumber.js';
import { getSkippedBlocksByEpoch, } from '../../actions/public/getSkippedBlocksByEpoch.js';
import { GetSponsorInfo, } from '../../actions/public/getSponsorInfo.js';
import { getStakingBalance, } from '../../actions/public/getStakingBalance.js';
import { getStatus, } from '../../actions/public/getStatus.js';
import { GetStorageAt, } from '../../actions/public/getStorageAt.js';
import { GetStorageRoot, } from '../../actions/public/getStorageRoot.js';
import { getSupplyInfo, } from '../../actions/public/getSupplyInfo.js';
import { getTransaction, } from '../../actions/public/getTransaction.js';
import { getTransactionReceipt, } from '../../actions/public/getTransactionReceipt.js';
import { getTxPoolNextNonce, } from '../../actions/public/getTxPoolNextNonce.js';
import { getVoteList, } from '../../actions/public/getVoteList.js';
import { multicall, } from '../../actions/public/multicall.js';
import { readContract, } from '../../actions/public/readContract.js';
import { simulateContract, } from '../../actions/public/simulateContract.js';
import { traceBlock, } from '../../actions/public/traceBlock.js';
import { traceTransaction, } from '../../actions/public/traceTransaction.js';
import { uninstallFilter, } from '../../actions/public/uninstallFilter.js';
import { verifyMessage, } from '../../actions/public/verifyMessage.js';
import { waitForTransactionReceipt, } from '../../actions/public/waitForTransactionReceipt.js';
import { watchEpochNumber, } from '../../actions/public/watchEpochNumber.js';
import { sendRawTransaction, } from '../../actions/wallet/sendRawTransaction.js';
export function publicActions(client) {
    return {
        getTransaction: (args) => getTransaction(client, args),
        getBlock: (args) => getBlock(client, args),
        getBastBlockHash: () => getBastBlockHash(client),
        getEpochNumber: (args) => getEpochNumber(client, args),
        getGasPrice: () => getGasPrice(client),
        estimateMaxPriorityFeePerGas: () => estimateMaxPriorityFeePerGas(client),
        getFeeHistory: (args) => getFeeHistory(client, args),
        getBlocksByEpoch: (args) => getBlocksByEpoch(client, args),
        getBalance: (args) => getBalance(client, args),
        getStakingBalance: (args) => getStakingBalance(client, args),
        getCollateralForStorage: (args) => getCollateralForStorage(client, args),
        getAdmin: (args) => getAdmin(client, args),
        getCode: (args) => getCode(client, args),
        getStorageAt: (args) => GetStorageAt(client, args),
        getStorageRoot: (args) => GetStorageRoot(client, args),
        getSponsorInfo: (args) => GetSponsorInfo(client, args),
        getNextNonce: (args) => getNextNonce(client, args),
        call: (args) => call(client, args),
        estimateGasAndCollateral: (args) => estimateGasAndCollateral(client, args),
        getLogs: (args) => getLogs(client, args),
        getTransactionReceipt: (args) => getTransactionReceipt(client, args),
        getAccount: (args) => getAccount(client, args),
        getInterestRate: (args) => getInterestRate(client, args),
        getAccumulateInterestRate: (args) => getAccumulateInterestRate(client, args),
        checkBalanceAgainstTransaction: (args) => checkBalanceAgainstTransaction(client, args),
        getSkippedBlocksByEpoch: (args) => getSkippedBlocksByEpoch(client, args),
        getConfirmationRiskByHash: (args) => getConfirmationRiskByHash(client, args),
        getStatus: () => getStatus(client),
        getClientVersion: () => getClientVersion(client),
        getBlockRewardInfo: (args) => getBlockRewardInfo(client, args),
        getBlockByHashWithPivotAssumption: (args) => getBlockByHashWithPivotAssumption(client, args),
        getDepositList: (args) => getDepositList(client, args),
        getVoteList: (args) => getVoteList(client, args),
        getSupplyInfo: () => getSupplyInfo(client),
        getAccountPendingInfo: (args) => getAccountPendingInfo(client, args),
        getAccountPendingTransactions: (args) => getAccountPendingTransactions(client, args),
        getPoSEconomics: (args) => getPoSEconomics(client, args),
        getPoSRewardByEpoch: (args) => getPoSRewardByEpoch(client, args),
        getParamsFromVote: (args) => getParamsFromVote(client, args),
        createEventFilter: (args) => createEventFilter(client, args),
        createBlockFilter: () => createBlockFilter(client),
        createPendingTransactionFilter: () => createPendingTransactionFilter(client),
        getFilterChanges: (args) => getFilterChanges(client, args),
        getFilterLogs: (args) => getFilterLogs(client, args),
        uninstallFilter: (args) => uninstallFilter(client, args),
        getCollateralInfo: (args) => getCollateralInfo(client, args),
        getTxPoolNextNonce: (args) => getTxPoolNextNonce(client, args),
        getPoSStatus: () => getPoSStatus(client),
        getPoSAccount: (args) => getPoSAccount(client, args),
        getPoSCommittee: (args) => getPoSCommittee(client, args),
        getPoSBlock: (args) => getPoSBlock(client, args),
        getPoSRewards: (args) => getPoSRewards(client, args),
        getPoSTransactionByNumber: (args) => getPoSTransactionByNumber(client, args),
        getEpochReceipts: (args) => getEpochReceipts(client, args),
        traceBlock: (args) => traceBlock(client, args),
        traceTransaction: (args) => traceTransaction(client, args),
        getChainId: () => getChainId(client),
        sendRawTransaction: (args) => sendRawTransaction(client, args),
        getFeeBurnt: () => getFeeBurnt(client),
        createContractEventFilter: (args) => createContractEventFilter(client, args),
        readContract: (args) => readContract(client, args),
        simulateContract: (args) => simulateContract(client, args),
        getContractEvents: (args) => getContractEvents(client, args),
        estimateContractGasAndCollateral: (args) => estimateContractGasAndCollateral(client, args),
        watchEpochNumber: (args) => watchEpochNumber(client, args),
        waitForTransactionReceipt: (args) => waitForTransactionReceipt(client, args),
        multicall: (args) => multicall(client, args),
        verifyMessage: (args) => verifyMessage(client, args),
        verifyTypedData: (args) => verifyTypedData(client, args),
    };
}
