// cive/_esm/clients/decorators/test.js
import { getCurrentSyncPhase, } from '../../actions/index.js';
import { clearTxpool } from '../../actions/test/clearTxpool.js';
import { createLocalNodeAccount, } from '../../actions/test/createLocalNodeAccount.js';
import { generateEmptyLocalNodeBlocks, } from '../../actions/test/generateEmptyLocalNodeBlocks.js';
import { generateLocalNodeBlock, } from '../../actions/test/generateLocalNodeBlock.js';
import { getLocalNodeAddresses, } from '../../actions/test/getLocalNodeAddresses.js';
import { lockLocalNodeAccount, } from '../../actions/test/lockLocalNodeAccount.js';
import { mine } from '../../actions/test/mine.js';
import { unlockLocalNodeAccount, } from '../../actions/test/unlockLocalNodeAccount.js';
export function testActions(client) {
    return {
        clearTxpool: () => clearTxpool(client),
        getLocalNodeAddresses: () => getLocalNodeAddresses(client),
        createLocalNodeAccount: (args) => createLocalNodeAccount(client, args),
        unlockLocalNodeAccount: (args) => unlockLocalNodeAccount(client, args),
        lockLocalNodeAccount: (args) => lockLocalNodeAccount(client, args),
        generateLocalNodeBlock: (args) => generateLocalNodeBlock(client, args),
        generateEmptyLocalNodeBlocks: (args) => generateEmptyLocalNodeBlocks(client, args),
        mine: (args) => mine(client, args),
        getCurrentSyncPhase: () => getCurrentSyncPhase(client),
    };
}
